utils/promis.js

```diff
@@ -0,0 +1,578 @@
+
+    const collection = 'promis_v1';
+    const doc = await surveyExists(collection, uid);
+    const surveyResults = doc.data();
+
+    const forms = Object.keys(promisConfig);
+    const scoresPayload = {};
+
+    for (let form of forms) {
+        const sourceQuestion = surveyResults[promisConfig[form].source];
+
+        if (sourceQuestion) {
+
+            let scoringData = {};
+
+            const questions = Object.keys(promisConfig[form].questions);
+
+            for (const question of questions) {
+                if (sourceQuestion[question]) {
+
+                    const questionConfig = promisConfig[form].questions[question];
+                    const questionResponse = sourceQuestion[question];
+
+                    scoringData[questionConfig.name] = questionConfig.responses[questionResponse];
+                }
+            }
+
+            const scores = await getScoringData(promisConfig[form].id, scoringData);
+
+            if (scores) {
+                scoresPayload[promisConfig[form].score] = parseInt(scores['T-Score']);
+                scoresPayload[promisConfig[form].error] = parseInt(scores['SError']);
+            }
+        }
+    }
+
+    if (Object.keys(scoresPayload).length > 0) {
+        await updateSurvey(scoresPayload, collection, doc);
+    }
+}
+
+const getScoringData = async (id, data) => {
+
+    const formData = new URLSearchParams();
+    const url = `https://www.assessmentcenter.net/ac_api/2013-01/Scores/`;
+
+    Object.keys(data).forEach(key => {
+        formData.append(key, data[key]);
+    });
+
+    try {
+        const response = await fetch(`${url}${id}.json`, {
+            method: "POST",
+            headers: {
+                "Authorization": "Basic " + generatePromisAuthToken(),
+                "Content-Type": "application/x-www-form-urlencoded"
+            },
+            body: formData.toString()
+        });
+
+        const scores = await response.json();
+
+        if (scores.ItemErrors) {
+            throw new Error('Errors in request data sent to PROMIS Scoring API', scores.ItemErrors);
+        }
+
+        return scores.Form[0];
+    }
+    catch (error) {
+        console.error('Error fetching PROMIS Scoring Data: ', error);
+        return null;
+    }
+}
+
+const promisConfig = {
+    'Physical Function': {
+        id: '7AA4002D-61C1-438D-9063-91A17FFA68BB',
+        source: 'D_284353934',
+        questions: {
+            'D_559540891': {
+                name: 'PFA11',
+                responses: {
+                    '367964536': 'Without any difficulty',
+                    '193058048': 'With a little difficulty',
+                    '263412634': 'With some difficulty',
+                    '734992634': 'With much difficulty',
+                    '221192556': 'Unable to do'
+                }
+            }, 
+            'D_917425212': {
+                name: 'PFA21',
+                responses: {
```

Comment:
Will it be better to save common responses as varaibles and refer to them?

Reply:
Let me try to think how I could minimize this... There were very subtle differences here and there in the responses.